### ISSUE TITLE
Update the syntax of Autohotkey

### ISSUE THREAD

**Issue by thqby** [github/linguist#6772]
## Describe the enhancement

AutoHotkey v2 has been officially released for more than a year, and it is the main version at present. In this month, v1 has stopped maintenance. AutoHotkey in github still uses the old syntax for language highlighting, so it's time to update it.

## URL of syntax highlighting grammar
https://github.com/thqby/vscode-autohotkey2-lsp/blob/main/syntaxes/ahk2.tmLanguage.json

## Most popular extensions
https://github.com/thqby/vscode-autohotkey2-lsp

**Comment by Banaanae** [comment]
I'm willing to work on this

**Comment by lildude** [comment]
https://github.com/github-linguist/linguist/pull/6815 is already in progress

**Comment by Banaanae** [comment]
I've look into this a bit more, and aren't too sure what to do.

While AutoHotkey has been released on V2, there are still many scripts using (the now deprecated) AutoHotkey V1. So I'm unsure whether we just update to V2 grammar (meaning _deprecated_ V1 files may lose some highlighting), or if there is a solution to have multiple grammar files based on a RegEx (something like heuristics?)

If that second option isn't possible forcing V2 as the highlighted option would be the way forward?

**Comment by thqby** [comment]
https://github.com/AutoHotkey/AutoHotkeyUX/blob/main/launcher.ahk#L62-L81

The ahk launcher uses regular expressions to differentiate versions and start them. How to act as a heuristic (a regex)?

**Comment by thqby** [comment]
I don't think heuristic is suitable for ahk. Only a small amount of code can be recognized by a single regular match.

The ahk launcher identifies the version by the difference in the number of features of the syntax of the two versions in the current code.

**Comment by Banaanae** [comment]
I think we can use [multiple RegExs](https://github.com/github-linguist/linguist/blob/master/lib/linguist/heuristics.yml#L426-L429)

> The ahk launcher identifies the version by the difference in the number of features of the syntax of the two versions in the current code.
But sometimes this detection fails, is there anyway to set a default language?

**Comment by thqby** [comment]
Some features are misidentified if they occur in a multiline string.
```
multiline_string := "
(
  #Require AutoHotkey v1
)"
```

Or some of the features of v1 are valid in the expression of v2.
ahk v1 command syntax `MsgBox, hello`
ahk v2 expression syntax
```
hello := 'hello'
user_func(
  MsgBox, hello
)
```

**Comment by DecimalTurn** [comment]
> But sometimes this detection fails, is there anyway to set a default language?

@Banaanae Yes, you could have something like this in the `heuristics.yml` file, so that if the regex doesn't match, for v2, it defaults to v1.

```
- extensions: ['.ahk']
  rules:
  - language: AutoHotKey v2
    pattern: '<your regex here>'
  - language: AutoHotKey v1
  
  ```

Also, you might want to use a `named_pattern` if you need multiple regex expressions as mentioned above. You'll find examples for this at the bottom of the `heuristics.yml` file.

**Comment by DecimalTurn** [comment]
Regarding whether or not V2 has reached a level of popularity that warrants a separate language, we're already at more than 2.9k files on GitHub using the `.ahk` extension that were written with AutoHotkey v2 :
https://github.com/search?q=path%3A*.ahk+%22%23Requires+AutoHotkey+v2.0%22+NOT+is%3Afork&type=code
